File: admin/src/api/erp_outbound_order.ts
import { PaginatedRequest, PaginatedResponse } from '@/base/page';
import { api } from '@/utils/request';
import { ErpOutboundOrderDetailBaseResponse, ErpOutboundOrderDetailRequest } from './erp_outbound_order_detail';
import { ErpOutboundOrderAttachmentBaseResponse, ErpOutboundOrderAttachmentRequest } from './erp_outbound_order_attachment';

const apis = {
  create_sale: '/erp/erp_outbound_order/create_sale', // 新增
  create_other: '/erp/erp_outbound_order/create_other', // 新增
  update_sale: '/erp/erp_outbound_order/update_sale', // 修改
  update_other: '/erp/erp_outbound_order/update_other', // 修改
  delete: '/erp/erp_outbound_order/delete', // 删除
  get: '/erp/erp_outbound_order/get', // 单条查询
  get_base_sales: '/erp/erp_outbound_order/get_base_sales', // 单条查询
  get_base_other: '/erp/erp_outbound_order/get_base_other', // 单条查询
  get_info_sales: '/erp/erp_outbound_order/get_info_sales', // 单条查询
  list: '/erp/erp_outbound_order/list', // 列表查询
  page_sales: '/erp/erp_outbound_order/page_sales', // 分页查询
  page_other: '/erp/erp_outbound_order/page_other', // 分页查询
}

export interface ErpOutboundOrderRequest {
  id?: number; // 出库订单ID
  sale_id?: number; // 销售订单ID
  customer_id?: number; // 客户ID
  outbound_date: string; // 出库日期
  remarks: string; // 备注
  discount_rate: number; // 优惠率（百分比，1000表示10.00%）
  other_cost: number; // 其他费用
  settlement_account_id?: number; // 结算账户ID

  details?: ErpOutboundOrderDetailRequest[]; // 入库采购产品仓库列表
  attachments: ErpOutboundOrderAttachmentRequest[]; // 入库附件列表
}

export interface ErpOutboundOrderResponse {
  id: number; // 出库订单ID
  order_number: number; // 订单编号
  sale_id: number; // 销售订单ID
  customer_id: number; // 客户ID
  user_id: number; // 用户ID
  outbound_date: string; // 出库日期
  remarks: string; // 备注
  discount_rate: number; // 优惠率（百分比，1000表示10.00%）
  other_cost: number; // 其他费用
  settlement_account_id: number; // 结算账户ID
  department_code: string; // 部门编码
  department_id: number; // 部门ID
  creator: number; // 创建者ID
  create_time: string; // 创建时间
  updater: number; // 更新者ID
  update_time: string; // 更新时间

  sales_order_number: number; // 销售订单编号
  customer_name: string; // 客户名
  settlement_account_name: string; // 结算账户

  details?: ErpOutboundOrderDetailBaseResponse[]; // 入库采购产品仓库列表
  attachments: ErpOutboundOrderAttachmentBaseResponse[]; // 入库附件列表
}

export interface ErpOutboundOrderQueryCondition extends PaginatedRequest {

}

export const createSaleErpOutboundOrder = (erp_outbound_order: ErpOutboundOrderRequest): Promise<number> => {
  return api.post<number>(apis.create_sale, erp_outbound_order);
}

export const createOtherErpOutboundOrder = (erp_outbound_order: ErpOutboundOrderRequest): Promise<number> => {
  return api.post<number>(apis.create_other, erp_outbound_order);
}

export const updateSaleErpOutboundOrder = (erp_outbound_order: ErpOutboundOrderRequest): Promise<void> => {
  return api.post<void>(apis.update_sale, erp_outbound_order);
}

export const updateOtherErpOutboundOrder = (erp_outbound_order: ErpOutboundOrderRequest): Promise<void> => {
  return api.post<void>(apis.update_other, erp_outbound_order);
}

export const deleteErpOutboundOrder = (id: number): Promise<void> => {
  return api.post<void>(`${apis.delete}/${id}`);
}

export const getErpOutboundOrder = (id: number): Promise<ErpOutboundOrderResponse> => {
  return api.get<ErpOutboundOrderResponse>(`${apis.get}/${id}`);
}

export const getBaseSaleErpOutboundOrder = (id: number): Promise<ErpOutboundOrderResponse> => {
  return api.get<ErpOutboundOrderResponse>(`${apis.get_base_sales}/${id}`);
}

export const getBaseOtherErpOutboundOrder = (id: number): Promise<ErpOutboundOrderResponse> => {
  return api.get<ErpOutboundOrderResponse>(`${apis.get_base_other}/${id}`);
}

export const getInfoSaleErpOutboundOrder = (id: number): Promise<ErpOutboundOrderResponse> => {
  return api.get<ErpOutboundOrderResponse>(`${apis.get_info_sales}/${id}`);
}

export const listErpOutboundOrder = (): Promise<Array<ErpOutboundOrderResponse>> => {
  return api.get<Array<ErpOutboundOrderResponse>>(apis.list);
}

export const pageSaleErpOutboundOrder = (condition: ErpOutboundOrderQueryCondition): Promise<PaginatedResponse<ErpOutboundOrderResponse>> => {
  return api.get<PaginatedResponse<ErpOutboundOrderResponse>>(apis.page_sales, condition);
}

export const pageOtherErpOutboundOrder = (condition: ErpOutboundOrderQueryCondition): Promise<PaginatedResponse<ErpOutboundOrderResponse>> => {
  return api.get<PaginatedResponse<ErpOutboundOrderResponse>>(apis.page_other, condition);
}

File: server/module/module-erp/erp-model/src/response/erp_outbound_order.rs
use chrono::NaiveDateTime;
use serde::{Serialize, Deserialize};
use utoipa::ToSchema;
use serde_with::{serde_as, DisplayFromStr};
use common::formatter::string_date_time::StringDateTime;

use crate::response::{erp_outbound_order_attachment::ErpOutboundOrderAttachmentBaseResponse, erp_outbound_order_detail::{ErpOutboundOrderDetailBaseOtherResponse, ErpOutboundOrderDetailBaseSalesResponse}};

#[serde_as]
#[derive(Deserialize, Serialize, Debug, Clone, ToSchema)]
pub struct ErpOutboundOrderResponse {
    
    pub id: i64, // 出库订单ID

    pub order_number: i64, // 订单编号
    
    pub sale_id: Option<i64>, // 销售订单ID
    
    pub customer_id: i64, // 客户ID

    pub user_id: i64, // 用户ID
    
    // #[serde_as(as = "DisplayFromStr")]
    // #[serde(with = "serde_with::chrono::naive_datetime")]
    #[serde_as(as = "common::formatter::string_date_time::StringDateTime")]
    #[schema(value_type = String, format = Date)]
    pub outbound_date: NaiveDateTime, // 出库日期
    
    pub remarks: Option<String>, // 备注
    
    pub discount_rate: Option<i64>, // 优惠率（百分比，1000表示10.00%）
    
    pub other_cost: Option<i64>, // 其他费用
    
    pub settlement_account_id: Option<i64>, // 结算账户ID
    
    pub department_code: String, // 部门编码
    
    pub department_id: i64, // 部门ID
    
    pub creator: Option<i64>, // 创建者ID
    
    // #[serde_as(as = "DisplayFromStr")]
    // #[serde(with = "serde_with::chrono::naive_datetime")]
    #[serde_as(as = "common::formatter::string_date_time::StringDateTime")]
    #[schema(value_type = String, format = Date)]
    pub create_time: NaiveDateTime, // 创建时间
    
    pub updater: Option<i64>, // 更新者ID
    
    // #[serde_as(as = "DisplayFromStr")]
    // #[serde(with = "serde_with::chrono::naive_datetime")]
    #[serde_as(as = "common::formatter::string_date_time::StringDateTime")]
    #[schema(value_type = String, format = Date)]
    pub update_time: NaiveDateTime, // 更新时间
    
}

#[serde_as]
#[derive(Deserialize, Serialize, Debug, Clone, ToSchema)]
pub struct ErpOutboundOrderPageSalesResponse {
    
    pub id: i64, // 出库订单ID

    pub order_number: i64, // 订单编号
    
    pub sale_id: Option<i64>, // 销售订单ID
    
    pub customer_id: i64, // 客户ID

    pub user_id: i64, // 用户ID
    
    // #[serde_as(as = "DisplayFromStr")]
    // #[serde(with = "serde_with::chrono::naive_datetime")]
    #[serde_as(as = "common::formatter::string_date_time::StringDateTime")]
    #[schema(value_type = String, format = Date)]
    pub outbound_date: NaiveDateTime, // 出库日期
    
    pub remarks: Option<String>, // 备注
    
    pub discount_rate: Option<i64>, // 优惠率（百分比，1000表示10.00%）
    
    pub other_cost: Option<i64>, // 其他费用
    
    pub settlement_account_id: Option<i64>, // 结算账户ID
    
    pub department_code: String, // 部门编码
    
    pub department_id: i64, // 部门ID
    
    pub creator: Option<i64>, // 创建者ID
    
    // #[serde_as(as = "DisplayFromStr")]
    // #[serde(with = "serde_with::chrono::naive_datetime")]
    #[serde_as(as = "common::formatter::string_date_time::StringDateTime")]
    #[schema(value_type = String, format = Date)]
    pub create_time: NaiveDateTime, // 创建时间
    
    pub updater: Option<i64>, // 更新者ID
    
    // #[serde_as(as = "DisplayFromStr")]
    // #[serde(with = "serde_with::chrono::naive_datetime")]
    #[serde_as(as = "common::formatter::string_date_time::StringDateTime")]
    #[schema(value_type = String, format = Date)]
    pub update_time: NaiveDateTime, // 更新时间

    /****************** 信息 ******************/

    pub sale_order_number: Option<i64>, // 销售订单编号

    pub customer_name: Option<String>, // 客户名

    pub settlement_account_name: Option<String>, // 结算账户名
    
}

#[serde_as]
#[derive(Deserialize, Serialize, Debug, Clone, ToSchema)]
pub struct ErpOutboundOrderPageOtherResponse {
    
    pub id: i64, // 出库订单ID

    pub order_number: i64, // 订单编号
    
    pub sale_id: Option<i64>, // 销售订单ID
    
    pub customer_id: i64, // 客户ID

    pub user_id: i64, // 用户ID
    
    // #[serde_as(as = "DisplayFromStr")]
    // #[serde(with = "serde_with::chrono::naive_datetime")]
    #[serde_as(as = "common::formatter::string_date_time::StringDateTime")]
    #[schema(value_type = String, format = Date)]
    pub outbound_date: NaiveDateTime, // 出库日期
    
    pub remarks: Option<String>, // 备注
    
    pub discount_rate: Option<i64>, // 优惠率（百分比，1000表示10.00%）
    
    pub other_cost: Option<i64>, // 其他费用
    
    pub settlement_account_id: Option<i64>, // 结算账户ID
    
    pub department_code: String, // 部门编码
    
    pub department_id: i64, // 部门ID
    
    pub creator: Option<i64>, // 创建者ID
    
    // #[serde_as(as = "DisplayFromStr")]
    // #[serde(with = "serde_with::chrono::naive_datetime")]
    #[serde_as(as = "common::formatter::string_date_time::StringDateTime")]
    #[schema(value_type = String, format = Date)]
    pub create_time: NaiveDateTime, // 创建时间
    
    pub updater: Option<i64>, // 更新者ID
    
    // #[serde_as(as = "DisplayFromStr")]
    // #[serde(with = "serde_with::chrono::naive_datetime")]
    #[serde_as(as = "common::formatter::string_date_time::StringDateTime")]
    #[schema(value_type = String, format = Date)]
    pub update_time: NaiveDateTime, // 更新时间

    /****************** 信息 ******************/

    pub customer_name: Option<String>, // 客户名

    pub settlement_account_name: Option<String>, // 结算账户名
    
}

#[serde_as]
#[derive(Deserialize, Serialize, Debug, Clone, ToSchema)]
pub struct ErpOutboundOrderBaseSalesResponse {
    
    pub id: i64, // 出库订单ID

    pub order_number: i64, // 订单编号
    
    pub sale_id: Option<i64>, // 销售订单ID
    
    pub customer_id: i64, // 客户ID

    pub user_id: i64, // 用户ID
    
    // #[serde_as(as = "DisplayFromStr")]
    // #[serde(with = "serde_with::chrono::naive_datetime")]
    #[serde_as(as = "common::formatter::string_date_time::StringDateTime")]
    #[schema(value_type = String, format = Date)]
    pub outbound_date: NaiveDateTime, // 出库日期
    
    pub remarks: Option<String>, // 备注
    
    pub discount_rate: Option<i64>, // 优惠率（百分比，1000表示10.00%）
    
    pub other_cost: Option<i64>, // 其他费用
    
    pub settlement_account_id: Option<i64>, // 结算账户ID

    pub details: Vec<ErpOutboundOrderDetailBaseSalesResponse>, // 出库销售产品仓库列表

    pub attachments: Vec<ErpOutboundOrderAttachmentBaseResponse>, // 出库附件列表
    
}

#[serde_as]
#[derive(Deserialize, Serialize, Debug, Clone, ToSchema)]
pub struct ErpOutboundOrderBaseOtherResponse {
    
    pub id: i64, // 出库订单ID

    pub order_number: i64, // 订单编号
    
    pub sale_id: Option<i64>, // 销售订单ID
    
    pub customer_id: i64, // 客户ID

    pub user_id: i64, // 用户ID
    
    // #[serde_as(as = "DisplayFromStr")]
    // #[serde(with = "serde_with::chrono::naive_datetime")]
    #[serde_as(as = "common::formatter::string_date_time::StringDateTime")]
    #[schema(value_type = String, format = Date)]
    pub outbound_date: NaiveDateTime, // 出库日期
    
    pub remarks: Option<String>, // 备注
    
    pub discount_rate: Option<i64>, // 优惠率（百分比，1000表示10.00%）
    
    pub other_cost: Option<i64>, // 其他费用
    
    pub settlement_account_id: Option<i64>, // 结算账户ID

    pub details: Vec<ErpOutboundOrderDetailBaseOtherResponse>, // 出库销售产品仓库列表

    pub attachments: Vec<ErpOutboundOrderAttachmentBaseResponse>, // 出库附件列表
    
}

#[serde_as]
#[derive(Deserialize, Serialize, Debug, Clone, ToSchema)]
pub struct ErpOutboundOrderInfoSalesResponse {
    
    pub id: i64, // 出库订单ID

    pub order_number: i64, // 订单编号
    
    pub sale_id: Option<i64>, // 销售订单ID
    
    pub customer_id: i64, // 客户ID

    pub user_id: i64, // 用户ID
    
    // #[serde_as(as = "DisplayFromStr")]
    // #[serde(with = "serde_with::chrono::naive_datetime")]
    #[serde_as(as = "common::formatter::string_date_time::StringDateTime")]
    #[schema(value_type = String, format = Date)]
    pub outbound_date: NaiveDateTime, // 出库日期
    
    pub remarks: Option<String>, // 备注
    
    pub discount_rate: Option<i64>, // 优惠率（百分比，1000表示10.00%）
    
    pub other_cost: Option<i64>, // 其他费用
    
    pub settlement_account_id: Option<i64>, // 结算账户ID
    
    pub department_code: String, // 部门编码
    
    pub department_id: i64, // 部门ID
    
    pub creator: Option<i64>, // 创建者ID
    
    // #[serde_as(as = "DisplayFromStr")]
    // #[serde(with = "serde_with::chrono::naive_datetime")]
    #[serde_as(as = "common::formatter::string_date_time::StringDateTime")]
    #[schema(value_type = String, format = Date)]
    pub create_time: NaiveDateTime, // 创建时间
    
    pub updater: Option<i64>, // 更新者ID
    
    // #[serde_as(as = "DisplayFromStr")]
    // #[serde(with = "serde_with::chrono::naive_datetime")]
    #[serde_as(as = "common::formatter::string_date_time::StringDateTime")]
    #[schema(value_type = String, format = Date)]
    pub update_time: NaiveDateTime, // 更新时间

    pub details: Vec<ErpOutboundOrderDetailBaseSalesResponse>, // 出库销售产品仓库列表

    pub attachments: Vec<ErpOutboundOrderAttachmentBaseResponse>, // 出库附件列表

    /****************** 信息 ******************/

    pub settlement_account_name: Option<String>, // 结算账户名
    
}

#[serde_as]
#[derive(Deserialize, Serialize, Debug, Clone, ToSchema)]
pub struct ErpOutboundOrderInfoOtherResponse {
    
    pub id: i64, // 出库订单ID

    pub order_number: i64, // 订单编号
    
    pub sale_id: Option<i64>, // 销售订单ID
    
    pub customer_id: i64, // 客户ID

    pub user_id: i64, // 用户ID
    
    // #[serde_as(as = "DisplayFromStr")]
    // #[serde(with = "serde_with::chrono::naive_datetime")]
    #[serde_as(as = "common::formatter::string_date_time::StringDateTime")]
    #[schema(value_type = String, format = Date)]
    pub outbound_date: NaiveDateTime, // 出库日期
    
    pub remarks: Option<String>, // 备注
    
    pub discount_rate: Option<i64>, // 优惠率（百分比，1000表示10.00%）
    
    pub other_cost: Option<i64>, // 其他费用
    
    pub settlement_account_id: Option<i64>, // 结算账户ID

    /****************** 信息 ******************/

    pub customer_name: Option<String>, // 客户名

    pub settlement_account_name: Option<String>, // 结算账户名

    pub details: Vec<ErpOutboundOrderDetailBaseOtherResponse>, // 出库销售产品仓库列表

    pub attachments: Vec<ErpOutboundOrderAttachmentBaseResponse>, // 出库附件列表
    
}